programs/psa/psa_hash.c

```diff
@@ -0,0 +1,147 @@
+
+#include "mbedtls/build_info.h"
+#include "mbedtls/platform.h"
+
+#define HASH_ALG PSA_ALG_SHA_256
+
+#define TEST_SHA256_HASH {                                                 \
+        0x5a, 0x09, 0xe8, 0xfa, 0x9c, 0x77, 0x80, 0x7b, 0x24, 0xe9, 0x9c, 0x9c, \
+        0xf9, 0x99, 0xde, 0xbf, 0xad, 0x84, 0x41, 0xe2, 0x69, 0xeb, 0x96, 0x0e, \
+        0x20, 0x1f, 0x61, 0xfc, 0x3d, 0xe2, 0x0d, 0x5a                          \
+}
+
+const uint8_t mbedtls_test_sha256_hash[] = TEST_SHA256_HASH;
+
+const size_t mbedtls_test_sha256_hash_len =
+    sizeof(mbedtls_test_sha256_hash);
+
+#if !defined(MBEDTLS_PSA_CRYPTO_C) || !defined(PSA_WANT_ALG_SHA_256)
+int main(void)
+{
+    mbedtls_printf("MBEDTLS_PSA_CRYPTO_C and MBEDTLS_SHA256_C"
+                   "not defined.\r\n");
+    return EXIT_SUCCESS;
+}
+#else
+
+int main(void)
+{
+    uint8_t buf[] = "Hello World!";
+    psa_status_t status;
+    uint8_t hash[PSA_HASH_LENGTH(HASH_ALG)];
+    size_t hash_size;
+    psa_hash_operation_t sha256_psa = PSA_HASH_OPERATION_INIT;
+    psa_hash_operation_t cloned_sha256 = PSA_HASH_OPERATION_INIT;
+
+    mbedtls_printf("PSA Crypto API: SHA-256 example\n\n");
+
+    status = psa_crypto_init();
+    if (status != PSA_SUCCESS) {
+        mbedtls_printf("psa_crypto_init failed\n");
+        return EXIT_FAILURE;
+    }
+
+
+    /* Compute hash using multi-part operation */
+
+    status = psa_hash_setup(&sha256_psa, HASH_ALG);
+    if (status != PSA_SUCCESS) {
+        mbedtls_printf("psa_hash_setup failed\n");
+        return EXIT_FAILURE;
+    }
+
+    status = psa_hash_update(&sha256_psa, buf, sizeof(buf));
+    if (status != PSA_SUCCESS) {
+        mbedtls_printf("psa_hash_update failed\n");
+        return EXIT_FAILURE;
+    }
+
+    status = psa_hash_clone(&sha256_psa, &cloned_sha256);
+    if (status != PSA_SUCCESS) {
+        mbedtls_printf("PSA hash clone failed");
+        return EXIT_FAILURE;
+    }
+
+    status = psa_hash_finish(&sha256_psa, hash, sizeof(hash), &hash_size);
+    if (status != PSA_SUCCESS) {
+        mbedtls_printf("psa_hash_finish failed\n");
+        return EXIT_FAILURE;
+    }
+
+    status =
+        psa_hash_verify(&cloned_sha256, mbedtls_test_sha256_hash, mbedtls_test_sha256_hash_len);
+    if (status != PSA_SUCCESS) {
+        mbedtls_printf("psa_hash_verify failed\n");
+        return EXIT_FAILURE;
+    } else {
+        mbedtls_printf("Multi-part hash operation successful!\n");
+    }
+
+    /* Compute hash using one-shot function call */
+    memset(hash, 0, sizeof(hash));
+    hash_size = 0;
+
+    status = psa_hash_compute(HASH_ALG,
+                              buf, sizeof(buf),
+                              hash, sizeof(hash),
+                              &hash_size);
+    if (status != PSA_SUCCESS) {
+        mbedtls_printf("psa_hash_compute failed\n");
+        return EXIT_FAILURE;
+    }
+
+    for (size_t j = 0; j < mbedtls_test_sha256_hash_len; j++) {
+        if (hash[j] != mbedtls_test_sha256_hash[j]) {
```

Comment:
Since the algo is now configurable the comparison is going to fail for any algo size than sha256, so I assume either the person who changes it going to modify mbedtls_test_sha256_hash to point his test data, or we provide different test data for different algos, either way I think it shouldn't have sha256 in the name.